### ISSUE TITLE
[Feature] Package names in summary audit report

### ISSUE THREAD

**Issue by iflp** [IBM/audit-ci#64]
Hi, I think it would be useful to include a summary of affected packages in the audit report much like how `yarn audit` displays it. My issue is the audit summary displays too little information, but the full report displays too much. Thank you for considering, and more than happy to put in a PR if needed.

**Comment by binarykitchen** [comment]
yep agree

**Comment by quinnturner** [comment]
I am interested in displaying a more user-friendly version of the audits for both NPM and Yarn. Right now the full report displays the raw JSON object received from `{npm|yarn} audit --json`, the summary displays the `metadata` field for NPM and Yarn displays anything with the `auditSummary` field.

There are a few different approaches we could take:

* Just remove "unnecessary" fields in the existing JSON objects and continue to show the JSON objects.
* Find a way to pretty-print the JSON objects received into their native package-manager print-style (perhaps https://github.com/npm/npm-audit-report#readme for NPM, not sure for Yarn). Unfortunately, since we use the `--json` flag at audit-time, we are missing out of the regular stdout from NPM and Yarn that you get running `{npm|yarn} audit`
* Introduce a new way to pretty-print that both NPM and Yarn use, likely based on `npm-audit-report` for ease of implementation.

Thoughts?

**Comment by quinnturner** [comment]
CC: @dubbha from #58.

**Comment by dubbha** [comment]
First of all, the summary output didn't even exist a week ago, before https://github.com/IBM/audit-ci/issues/58 and https://github.com/IBM/audit-ci/pull/59

The use case I was covering is described in https://github.com/IBM/audit-ci/issues/58 and the current output covers exactly what I consider to be a summary:

```js
{
    "vulnerabilities": {
      "info": 0,
      "low": 9,
      "moderate": 213,
      "high": 4,
      "critical": 0
    },
    "dependencies": 70026,
    "devDependencies": 0,
    "optionalDependencies": 0,
    "totalDependencies": 70026
  }
```
It outputs this and yet doesn't fail because of the `--critical`. The idea behind this is: I don't what the whole CI to fail unless there are critical vulnerabilities, but I want a summary output containing a count for the lower-level vulnerabilities, so that I know when these appear and can fix them when I have time.

The output above is from a real project, and it is exactly sufficient for my use case. Now I don't really get the idea, what would you like it to look like instead? Since I only really care for criticals (remember the `--critical`) I don't need the output to be too polluted (hence the PR in the first place). And my guess is that with       `"low": 9`, `"moderate": 213`, `"high": 4`, any kind of "summary of affected packages" would pollute the log extensively.

**Comment by iflp** [comment]
Thanks for the input @dubbha.

I do not think a summary of affected packages would pollute the logs. Aside from the initial audit on a new project (and fix), there will not be 100s of advisories popping up everyday. There should only be a few at most, `npm audit` or `yarn audit` does a great job of showing just the essentials without being overly verbose.

I think there are merits to having a concise summary like what you have mentioned, and a pretty print version with just the essentials eg.:

```
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Prototype Pollution                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ hoek                                                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ > 4.2.0 < 5.0.0 || >= 5.0.3                                  │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ less                                                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ less > request > hawk > sntp > hoek                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/566                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
```

@quinnturner : Approach 3 sounds good, maybe with a `--package-summary` flag or something

**Comment by dubbha** [comment]
> Thanks for the input @dubbha.

You're welcome.

> I do not think a summary of affected packages would pollute the logs.

In fact it will pollute the logs.

> Aside from the initial audit on a new project (and fix), there will not be 100s of advisories popping up everyday.

You should not be assuming an initial fix of the issues below my threshold. If I am running `audit-ci --critical` I only really care for criticals. And you should not be assuming I will even bother to fix the other, lower-level issues, ever. All I need to know is if there are any, that's it. At least that's the idea behind a `--summary` flag.

What you're talking about is basically returning from a summary to the full output. Even if it's changed from JSON to stdout/pretty-print it's still going back to the full output. This case it makes sense to discuss chaning the `--report` output, but not the `--summary` output.

My use case above is 100% legit, and the output comes from a real project where there are no plans to address the minor issues in the nearest future. And with your approach the logs will (again) be polluted to the extent the CI output will become unusable (see below). It will output a "pretty-print" for each an every instance of the mentioned vulnerabilities ("low": 9, "moderate": 213, "high": 4) which I don't really care for in the first place. If I do care for them later, I will run the audit locally and fix them.

Here's how the output would look like:

LINES 1 to 31:
```
yarn audit
yarn audit v1.13.0
warning package.json: License should be a valid SPDX license expression
warning @linsi/templateengine@8.0.0: License should be a valid SPDX license expression
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ low           │ Regular Expression Denial of Service                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ braces                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=2.3.1                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ ts-jest                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ ts-jest > cpx > chokidar > anymatch > micromatch > braces    │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/786                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Prototype Pollution                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ lodash                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=4.17.11                                                    │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ ts-jest                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ ts-jest > babel-plugin-istanbul > istanbul-lib-instrument >  │
│               │ babel-generator > babel-types > lodash                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/782                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
```

... SO MANY LINES ...

LINES 3196 - 3225
```
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ high          │ Denial of Service                                            │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ https-proxy-agent                                            │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=2.2.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ typings                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ typings > typings-core > popsicle-proxy-agent >              │
│               │ https-proxy-agent                                            │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/593                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Prototype Pollution                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ lodash                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=4.17.11                                                    │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ @linsi/entities                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ @linsi/entities > lodash                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/782                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
226 vulnerabilities found - Packages audited: 70027
Severity: 9 Low | 213 Moderate | 4 High
Done in 2.63s.
```
In my particular use case having 3225 lines of the info I don't really care for at the point is a pollution.

Also it would weigh 527K in size alone, and together with the other logs sooner or later would hit the CI logs sizing limitation like the one mentioned in the original https://github.com/IBM/audit-ci/issues/58

```
Job's log exceeded limit of 4194304 bytes.
```

So your pretty-print idea contradicts the use case `--summary` was introduced for. Either consider dicussing changing the `--report` output, or maybe introduce your own flag.

**Comment by quinnturner** [comment]
@dubbha @iflp Great discussions all around, thanks! What do you think of introducing another report variant which shows both the summary _and_ the vulnerabilties that are breaking the build? Currently, `--report` shows _all_ vulnerabilities including `low` when you use the `--critical` flag. `--summary` just shows the `auditSummary`.

The new flag (unknown name) would show the summary and would show the output only for the vulnerabilities that break the CI build (critical in this case) as well as the summary. Alternatively, we could repurpose `--summary` to do this, as this may be the desired behaviour.

**Comment by quinnturner** [comment]
Hi everyone!

I have created #74. Please consider reviewing!

**Comment by quinnturner** [comment]
Note that I revised the potential approach in a new comment: https://github.com/IBM/audit-ci/pull/74#issuecomment-479696523

Exerpt:

>One alternative that I am interested in as well would be adding the flag --report-type {summary, important, full}. This set up could support different report types in the future. I suggest we would default to important, which is the behaviour in the current PR. We would continue by deprecating the existing --summary and --report flags but continue to support them for "a while", piping them to the associated --report-type (summary --> summary and report --> full) and console.warn the deprecation and potential future removal.

>This approach would be a major bump to the version and would include documentation in the README for upgrading an upgrading strategy.

Open to discussion and preferences!

**Comment by quinnturner** [comment]
I have released `v2.0.0` which addresses this issue by adding the `--report-type` option; which defaults to a new setting `important`. `important` shows the audit summary as well as _relevant_ vulnerabilities instead of the full audit report or no _relevant_ vulnerabilities. For now, I will close this. I see the value of adding a different way of formatting the output, but the original issue of adding package names in the summary audit report has been addressed. Feel free to open another issue for formatting the output if you feel that it would provide significant value.
